lib/lambda_ethereum_consensus/beacon/pending_blocks.ex

```diff
@@ -68,44 +71,39 @@ defmodule LambdaEthereumConsensus.Beacon.PendingBlocks do
   @impl true
   @spec handle_info(atom(), state()) :: {:noreply, state()}
   def handle_info(:process_blocks, state) do
-    state.pending_blocks
+    state
+    |> Map.filter(fn {_, {_, s}} -> s == :pending end)
+    |> Enum.map(fn {root, {block, _}} -> {root, block} end)
     |> Enum.sort_by(fn {_, signed_block} -> signed_block.message.slot end)
     |> Enum.reduce(state, fn {block_root, signed_block}, state ->
       parent_root = signed_block.message.parent_root
 
       cond do
+        # If already processed, remove it
+        Blocks.get_block(block_root) ->
+          state |> Map.delete(block_root)
+
         # If parent is invalid, block is invalid
-        state.invalid_blocks |> Map.has_key?(parent_root) ->
+        get_block_status(state, parent_root) == :invalid ->
+          state |> Map.put(block_root, {nil, :invalid})
+
+        # If parent is processing, block is pending
+        get_block_status(state, parent_root) == :processing ->
           state
-          |> Map.update!(:pending_blocks, &Map.delete(&1, block_root))
-          |> Map.update!(
-            :invalid_blocks,
-            &Map.put(&1, block_root, signed_block.message |> Map.take([:slot, :parent_root]))
-          )
 
         # If parent is pending, block is pending
-        state.pending_blocks |> Map.has_key?(parent_root) ->
+        get_block_status(state, parent_root) == :pending ->
           state
 
-        # If already in fork choice, remove from pending
-        ForkChoice.has_block?(block_root) ->
-          state |> Map.update!(:pending_blocks, &Map.delete(&1, block_root))
-
         # If parent is not in fork choice, download parent
-        not ForkChoice.has_block?(parent_root) ->
-          state |> Map.update!(:blocks_to_download, &MapSet.put(&1, parent_root))
+        !Blocks.get_block(parent_root) ->
```

Comment:
This introduces a bug: if the parent is in the DB this would be false, and we'd try to process the block even if its parent wasn't processed, causing it to fail and marking it as an invalid block

Reply:
But this assumes that a block is stored in the DB only if it is processed (and valid). We should not store invalid (or potentially invalid) blocks.

Do we store blocks before processing them today?

Reply:
No. But this can happen if the node restarts after processing some blocks

Reply:
I think that this is probably out of scope from this PR

Reply:
For sure. We could patch it up by just nuking the database in the makefile to avoid this error when debugging, but we need to find a better way to handle this.

Reply:
Would you like me to add this to the PR?

Reply:
I think this is a more complex discussion outside of the scope of the PR, which is supporting recoverability. To do so, we should save the state of a block in the DB.

Reply:
Just to clarify, I meant we need to find a better way to handle this _in a following PR_. For now, we can just go for the nuclear option.